src/WinRT.Runtime/ActivationFactory.cs

```diff
@@ -315,8 +366,37 @@ public static ObjectReference<I> Get<I>(string typeName, Guid iid)
             }
         }
 
+        private static void ThrowIfClassNotRegisteredAndManifestFreeActivationDisabled(string typeName, int hr)
+        {
+            // If manifest free activation is enabled, we never throw here.
+            // Callers will always try the fallback path if we didn't succeed.
+            if (FeatureSwitches.EnableManifestFreeActivation)
+            {
+                return;
+            }
+
+            [MethodImpl(MethodImplOptions.NoInlining)]
+            static Exception GetException(string typeName, int hr)
+            {
+                Exception exception = Marshal.GetExceptionForHR(hr)!;
+
+                if (hr == ExceptionHelpers.REGDB_E_CLASSNOTREG)
+                {
+                    throw new NotSupportedException(
+                        $"Failed to activate type with runtime class name '{typeName}' with 'RoGetActivationFactory' (it returned 0x80040154, ie. 'REGDB_E_CLASSNOTREG'). Make sure to add the activatable class id for the type " +
+                        "to the APPX manifest, or enable the manifest free activation fallback path by setting the 'CsWinRTEnableManifestFreeActivation' property (note: the fallback path incurs a performance hit).", exception);
+                }
+
+                return exception;
+            }
+
+            // Explicit throw so the JIT can see it and correctly optimize for always throwing paths.
+            // The exception instantiation is in a separate method so that codegen remains separate.
+            throw GetException(typeName, hr);
+        }
+
 #if NET
-        private static IObjectReference GetFromActivationHandler(string typeName, Guid iid)
+        private static IObjectReference? GetFromActivationHandler(string typeName, Guid iid)

```

Comment:
Just to confirm, adding nullable annotations here and above are not compat concerns right given nullable wasn't enabled before?

Reply:
Correct, this was just nullable oblivious before, so this is adding new info to it, but not breaking.